### PR TITLE
Metadata: Allow authenticated mongodb connections; Fix #6605

### LINKED ISSUE #6605: MongoDB metadata implementation does not allow using authentication
### Description

There is now way to configure username / password for the mongo db connection

### Motivation

Any real deployment will need authenticated access to a database

### Change

Add configuration options `mongo_service_user` and `mongo_service_password` in the `[metdata]` section, use to create and authenticated client.

I can open a PR

### PR DESCRIPTION
**Pull request by maxnoe**
Follow up to #6606 following the contributors guide

### CODE REVIEW

File: lib/rucio/core/did_meta_plugins/mongo_meta.py [hunk 1]
```diff
@@ -47,7 +55,16 @@ def __init__(self, host=None, port=None, db=None, collection=None):
         if collection is None:
             collection = config.config_get('metadata', 'mongo_collection')
 
-        self.client = pymongo.MongoClient("mongodb://{}:{}/".format(host, port))
+        if user is None:
+            user = config.config_get("metadata", "mongo_user", default=None)
+        if user is not None and password is None:
+            password = config.config_get("metadata", "mongo_password")
+        if user is not None:
+            auth = "{user}:{password}@".format(user=user, password=password)
```
Comment (rdimaio) [inline lib/rucio/core/did_meta_plugins/mongo_meta.py]: this could be `if user is not None and password is not None` to be extra sure that `password` exists

File: tests/test_did_meta_plugins.py [hunk 2]
```diff
@@ -198,7 +198,9 @@ def mongo_meta():
         host='mongo',
         port=27017,
         db='test_db',
-        collection='test_collection'
+        collection='test_collection',
+        user="rucio",
+        password="mongo-meta",
```
Comment (rdimaio) [inline tests/test_did_meta_plugins.py]: It would be great if we had another `mongo_meta` fixture without `user` and `password` (and corresponding tests with that fixture), to ensure that unauthenticated mongodb connections still work

File: lib/rucio/core/did_meta_plugins/mongo_meta.py (hunk 1 above)
Comment (maxnoe) [reply lib/rucio/core/did_meta_plugins/mongo_meta.py]: actually, it can just be merged with  the if above. If `password is None` and doesn't appears in the configuration, we error.

File: tests/test_did_meta_plugins.py (hunk 2 above)
Comment (maxnoe) [reply tests/test_did_meta_plugins.py]: That would need a second mongodb in the docker compose yaml I guess?

Comment (maxnoe) [reply tests/test_did_meta_plugins.py]: I added the second mongo definition and parametrized the test to test both authenticated and unauthenticated database connections

File: lib/rucio/core/did_meta_plugins/mongo_meta.py (hunk 1 above)
Comment (maxnoe) [reply lib/rucio/core/did_meta_plugins/mongo_meta.py]: done

File: AUTHORS.rst [hunk 3]
```diff
@@ -66,6 +66,7 @@ Individual contributors to the source code
 - Jensen Zhang <hack@jensen-zhang.site>, 2022
 - Aksel Lunde Aase <aksel.lunde.aase@gmail.com>, 2022
 - Elena Gazzarrini <gazzarrini.elena@gmail.com>, 2022-2023
+- Maximilian Linhoff, <maximilian.linhoff@tu-dortmund.de>, 2024
```
Comment (rdimaio) [inline AUTHORS.rst:69]: The changes to `AUTHORS.rst` should be in a separate commit

File: lib/rucio/core/did_meta_plugins/mongo_meta.py [hunk 4]
```diff
@@ -36,7 +36,15 @@
...
 class MongoDidMeta(DidMetaPlugin):
-    def __init__(self, host=None, port=None, db=None, collection=None):
+    def __init__(
+        self,
+        host: "Optional[str]" = None,
+        port: "Optional[int]" = None,
+        db: "Optional[str]" = None,
+        collection: "Optional[str]" = None,
+        user: "Optional[str]" = None,
+        password: "Optional[str]" = None,
```
Comment (rdimaio) [inline lib/rucio/core/did_meta_plugins/mongo_meta.py:46]: Since these type hints only use built-in types and `Optional` from `typing`, we could move `from typing import Optional` outside of the `if TYPE_CHECKING` block and remove the quotes from these type annotations.

I'm just leaving this as a small note, no need to make any changes, as this is something that we can lint automatically with Ruff. We temporarily disabled the code that handles this because of false positives: https://github.com/rucio/rucio/blob/fb42cda0c8c49ec11a3fe624be2ca17d208887a8/pyproject.toml#L59, but once we enable it again this will be formatted automatically.

**Review by rdimaio — COMMENTED** [review]
Ruff failures are due to an unrelated issue that was fixed in a separate PR (they will go away if you rebase, but there's no need to do that since there's no merge conflicts)

File: AUTHORS.rst (hunk 3 above)
Comment (maxnoe) [reply AUTHORS.rst:69]: Sure

**Review by rdimaio — APPROVED** [review]
